### PR TITLE
Fix httpx is not encoding with symbol '%s' (#3140)

### PR DESCRIPTION
**Pull request by Tunglies**
Fix httpx is not encoding with symbol '%s' (#3140)
Add url parsing test

# Summary

# Checklist

- [x] I understand that this PR may be closed in case there was no previous discussion. (This doesn't apply to typos!)
- [x] I've added a test for each change that was introduced, and I tried as much as possible to make a single atomic change.
- [ ] I've updated the documentation accordingly.

### CODE REVIEW

File: httpx/_urlparse.py [hunk 1]
```diff
@@ -424,13 +424,23 @@ def is_safe(string: str, safe: str = "/") -> bool:
...
+            char == "%"
+            and i + 2 < len(string)
+            and string[i + 1 : i + 3].isalnum()
+            and len(string[i + 1 : i + 3]) == 2
+        ):
+            i += 3
+        else:
             return False
-
     return True
```
Comment (T-256) [inline httpx/_urlparse.py:444]:
```suggestion

    # All characters must already be non-escaping or valid percent-encoded
    for index, char in enumerate(string):
        if char not in NON_ESCAPED_CHARS and not PERCENT_ENCODED_REGEX.match(
            string[index : index + 3]
        ):
            return False

    return True
```

Comment (Tunglies) [reply httpx/_urlparse.py:444]: Thanks for review! I have applied the suggestion.

**Review by T-256 — APPROVED** [review]
Thanks, I think it deserves a note in changelog.

**Comment by heysarthak** [comment]
Hi @T-256,
will the code change encode every "%" sent into uri as "%25" ?

**Comment by T-256** [comment]
> Hi @T-256, will the code change encode every "%" sent into uri as "%25" ?

Yes, if it is invalid percent-encoded.

**Comment by heysarthak** [comment]
when can we merge this changes? @tomchristie

**Comment by jhominal** [comment]
Hello, could you look at if this PR fixes the cases of bad `%` character encoding that are raised in https://github.com/encode/httpx/discussions/3135?

**Review by sileht — APPROVED** [review]

**Comment by tomchristie** [comment]
Thanks. I think this is superseeded by #3187.

**Comment by heysarthak** [comment]
I tried the same on http 0.27.0, still having the same issue.

**Comment by tomchristie** [comment]
Thanks everyone.
Remaining issues here should now all fully be addressed in #3371 and #3373.

**Pull request closed without merging**